File: CIMenu/DataFetcher.rb
class DataFetcher
  URL = "https://semaphoreapp.com/api/v1/projects?auth_token="
  #URL = "http://localhost/projects.json?auth_token="
  INTERVAL = 60

  def initialize
    NSNotificationCenter.defaultCenter.addObserver(self,
      selector:"fetch:",
      name:"com.cimenu.CIMenu.preferences.token.changed",
      object:nil)
  end

  def fetch(timer = nil)
    request = NSMutableURLRequest.requestWithURL(url)

    NSURLConnection.connectionWithRequest(request, delegate:ConnectionDelegate.new)

    # connection = NSURLConnection.alloc.initWithRequest request,
    #   delegate:ConnectionDelegate.new,
    #   startImmediately:false
    # connection.scheduleInRunLoop NSRunLoop.mainRunLoop, forMode:NSRunLoopCommonModes
    # connection.start
  end

  def startTimer
    @timer = NSTimer.scheduledTimerWithTimeInterval INTERVAL,
      target:self,
      selector:'fetch:',
      userInfo:nil,
      repeats:true
    # NSRunLoop.mainRunLoop.addTimer @timer, forMode:NSRunLoopCommonModes
  end

  def stopTimer
    @timer.invalidate && @timer = nil if @timer
  end

  private

  def apiKey
    defaults = NSUserDefaults.standardUserDefaults
    token = defaults.objectForKey 'org.cimenu.apikey'
    p "fetching with #{token}"
    token
  end

  def url
    NSURL.URLWithString("#{URL}#{apiKey}")
  end
end

File: CIMenu/DataFetcher.swift
//
//  DataFetcher.swift
//  CIMenu
//
//  Created by Nickolay Abdrafikov on 28.08.14.
//  Copyright (c) 2014 Nickolay Abdrafikov. All rights reserved.
//

class DataFetcher {
    let url = "https://semaphoreapp.com/api/v1/projects?auth_token="
//    let url = "http://localhost/projects.json?auth_token="
    let interval = NSTimeInterval(15)
    let defaults = NSUserDefaults.standardUserDefaults()
    let notificationCenter = NSNotificationCenter.defaultCenter()

    var timer: NSTimer!

    init() {
//        notificationCenter.addObserver(self,
//            selector: Selector("fetch"),
//            name: "org.cimenu.preferences.token.changed",
//            object: nil
//        )
    }

    // without @objc 'Unrecognized selector' exeptions raises
    @objc func fetch() {
        let token = defaults.objectForKey("org.cimenu.apikey") as String

        self.notificationCenter.postNotificationName(
            "org.cimenu.semaphore.connectionWillStartLoading",
            object: nil,
            userInfo: nil
        )
        println("GET \(url + token)")
        request(.GET, url + token)
            .responseString { (request, response, string, error) in
                let json = JSON.parse(string!)

                self.notificationCenter.postNotificationName(
                    "org.cimenu.semaphore.dataReceived",
                    object: nil,
                    userInfo: ["json" : json]
                )
            }

    }

    func startTimer() {
        timer = NSTimer.scheduledTimerWithTimeInterval(interval,
            target: self,
            selector: Selector("fetch"),
            userInfo: nil,
            repeats: true
        )
    }

    func stopTimer() {
        timer?.invalidate
        timer = nil
    }

    deinit {
        stopTimer()
    }
}
